File: core-services/internal/cronjobs/workreaper/workreaper.go
// The workreaper package implements a cron job to delete old work-items and workflow steps
package workreaper

import (
	"context"
	"fmt"

	"github.com/nasa/harmony/core-services/internal/db"
	logs "github.com/nasa/harmony/core-services/internal/log"
	"github.com/nasa/harmony/core-services/internal/models/job"
)

func deleteTerminalWorkItems(ctx context.Context, notUpdatedForMinutes int, jobStatus []job.JobStatus, batchSize int) {
	var done = false
	var startingId = 0
	var totalDeleted int64 = 0

	logger := logs.GetLoggerForContext(ctx)

	logger.Info("Work reaper delete terminal items started")

	for !done {
		workItemIds, err := db.GetWorkItemIdsByJobUpdateAgeAndStatus(ctx, notUpdatedForMinutes, jobStatus, startingId, batchSize)
		if err != nil {
			logger.Error("Failed to get work-item ids for deletion")
		}
		if len(workItemIds) > 0 {
			logger.Info(fmt.Sprintf("Deleting %d work-items", len(workItemIds)))
			numDeleted, err := db.DeleteWorkItemsById(ctx, workItemIds)
			if err != nil {
				logger.Error("Failed to delete work-items")
			}
			totalDeleted += numDeleted
		}

		if len(workItemIds) < batchSize {
			done = true
		}
	}

	logger.Info(fmt.Sprintf("Done deleting work-items. Total work-items deleted: %d", totalDeleted))
}

func deleteTerminalWorkflowSteps(ctx context.Context, notUpdatedForMinutes int, jobStatus []job.JobStatus, batchSize int) {
	var done = false
	var startingId = 0
	var totalDeleted int64 = 0

	logger := logs.GetLoggerForContext(ctx)

	logger.Info("Work reaper delete terminal workflow steps started")

	for !done {
		workItemIds, err := db.GetWorkflowStepIdsByJobUpdateAgeAndStatus(ctx, notUpdatedForMinutes, jobStatus, startingId, batchSize)
		if err != nil {
			logger.Error("Failed to get workflow step ids for deletion")
		}
		if len(workItemIds) > 0 {
			logger.Info(fmt.Sprintf("Deleting %d workflow steps", len(workItemIds)))
			numDeleted, err := db.DeleteWorkflowStepsById(ctx, workItemIds)
			if err != nil {
				logger.Error("Failed to delete workflow steps")
			}
			totalDeleted += numDeleted
		}

		if len(workItemIds) < batchSize {
			done = true
		}
	}

	logger.Info(fmt.Sprintf("Done deleting workflow steps. Total workflow steps deleted: %d", totalDeleted))
}

func DeleteOldWork(ctx context.Context, env WorkReaperEnv) {
	deleteTerminalWorkItems(ctx, env.ReapableWorkAgeMinutes, job.TerminalStatuses, env.WorkReaperBatchSize)
	deleteTerminalWorkflowSteps(ctx, env.ReapableWorkAgeMinutes, job.TerminalStatuses, env.WorkReaperBatchSize)

}

File: services/cron-service/app/cronjobs/work-reaper.ts
import { JobStatus, terminalStates } from '../../../harmony/app/models/job';
import {
  deleteWorkItemsById, getWorkItemIdsByJobUpdateAgeAndStatus,
} from '../../../harmony/app/models/work-item';
import {
  deleteWorkflowStepsById, getWorkflowStepIdsByJobUpdateAgeAndStatus,
} from '../../../harmony/app/models/workflow-steps';
import { Context } from '../util/context';
import env from '../util/env';
import { CronJob } from './cronjob';

/**
 * Find work items that are older than notUpdatedForMinutes and delete them.
 * @param notUpdatedForMinutes - upper limit on the duration since the last update
 * @param jobStatus - a list of terminal job statuses
 * @returns Resolves when the request is complete
 */
async function deleteTerminalWorkItems(ctx: Context, notUpdatedForMinutes: number, jobStatus: JobStatus[]): Promise < void> {
  let done = false;
  let startingId = 0;
  let totalDeleted = 0;
  const batchSize = env.workReaperBatchSize;
  const { logger, db } = ctx;
  logger.info('Work reaper delete terminal work items started.');

  while (!done) {
    try {
      const workItemIds = await getWorkItemIdsByJobUpdateAgeAndStatus(
        db, notUpdatedForMinutes, jobStatus, startingId, batchSize,
      );
      if (workItemIds.length > 0) {
        const numItemsDeleted = await deleteWorkItemsById(db, workItemIds);
        totalDeleted += numItemsDeleted;
        logger.info(`Work reaper removed ${numItemsDeleted} work items, starting id: ${startingId}.`);
        startingId = workItemIds[workItemIds.length - 1];
      } else {
        logger.info('Work reaper did not find any work items to delete');
      }

      if (workItemIds.length < batchSize) {
        done = true;
      }
    } catch (e) {
      done = true;
      logger.error('Error attempting to delete terminal work items');
      logger.error(e);
    }
  }
  logger.info(`Work reaper delete terminal work items completed. Total work items deleted: ${totalDeleted}`);
}


/**
 * Find workflow steps that are older than notUpdatedForMinutes and delete them.
 * @param notUpdatedForMinutes - upper limit on the duration since the last update
 * @param jobStatus - a list of terminal job statuses
 * @returns Resolves when the request is complete
 */
async function deleteTerminalWorkflowSteps(ctx: Context, notUpdatedForMinutes: number, jobStatus: JobStatus[]): Promise < void> {
  let done = false;
  let startingId = 0;
  let totalDeleted = 0;
  const batchSize = env.workReaperBatchSize;
  const { logger, db } = ctx;
  logger.info('Work reaper delete terminal workflow steps started.');

  while (!done) {
    try {
      const workflowSteps = await getWorkflowStepIdsByJobUpdateAgeAndStatus(
        db, notUpdatedForMinutes, jobStatus, startingId, batchSize,
      );
      if (workflowSteps.length > 0) {
        const numItemsDeleted = await deleteWorkflowStepsById(db, workflowSteps);
        totalDeleted += numItemsDeleted;
        logger.info(`Work reaper removed ${numItemsDeleted} workflow steps, starting id: ${startingId}.`);
        startingId = workflowSteps[workflowSteps.length - 1];
      } else {
        logger.info('Work reaper did not find any workflow steps to delete');
      }

      if (workflowSteps.length < batchSize) {
        done = true;
      }
    } catch (e) {
      done = true;
      logger.error('Error attempting to delete terminal workflow steps');
      logger.error(e);
    }
  }
  logger.info(`Work reaper delete terminal workflow steps completed. Total workflow steps deleted: ${totalDeleted}`);
}

/**
 * Work reaper class for cron service
 */
export class WorkReaper extends CronJob {

  static async run(ctx: Context): Promise<void> {
    const { logger } = ctx;
    logger.debug('Running');
    try {
      await deleteTerminalWorkItems(
        ctx,
        env.reapableWorkAgeMinutes,
        terminalStates,
      );
      await deleteTerminalWorkflowSteps(
        ctx,
        env.reapableWorkAgeMinutes,
        terminalStates,
      );
    } catch (e) {
      logger.error('Work reaper failed to delete terminal work');
      logger.error(e);
    }
  }
}